layout prepared

DemasX64 opened the pull request.

DemasX64 committed:
layout prepared

Zaytsev-Alex commented on src/App.tsx:
```diff
@@ -1,24 +1,18 @@
 import React from 'react';
 import logo from './logo.svg';
 import './App.css';
```
Нужно исправить: стили должны быть добавлены через css-modules, если они не являются глобальными.

Zaytsev-Alex replied:
Нужно исправить: стили следует подключать используя css-modules, если это не глобальные стили.

Zaytsev-Alex commented on src/App.tsx:
```diff
@@ -1,24 +1,18 @@
 import React from 'react';
```
Нужно исправить: компонент App вместе с его стилями следует перенести в папку компонентов -> папка app

Zaytsev-Alex commented on src/components/app-header/app-header.js:
```diff
@@ -0,0 +1,35 @@
+                    <nav className={styles.nav}>
+                        <NavItem title="Конструктор"value={1} active={this.state.current===1} onClick={this.setCurrent}><BurgerIcon type={this.state.current===1?"primary":'secondary'} /></NavItem>
+                        <NavItem title="Лента заказов" value={2} active={this.state.current===2} onClick={this.setCurrent}><ListIcon type={this.state.current===2?"primary":'secondary'} /></NavItem>
+                        <div style={{flex:1}}></div>
```
Нужно исправить: следует использовать модульные стили вместо прямого инлайнового задания.

Zaytsev-Alex commented on src/components/app-header/app-header.js:
```diff
@@ -0,0 +1,35 @@
+                    </div>
+                    <nav className={styles.nav}>
+                        <NavItem title="Конструктор"value={1} active={this.state.current===1} onClick={this.setCurrent}><BurgerIcon type={this.state.current===1?"primary":'secondary'} /></NavItem>
+                        <NavItem title="Лента заказов" value={2} active={this.state.current===2} onClick={this.setCurrent}><ListIcon type={this.state.current===2?"primary":'secondary'} /></NavItem>
```
Можно лучше: логику лучше выносить из ретёрна, в методе render над ретёрном можно задать константы определяющие статус, к примеру const isConstructorActive = this.state.current === 1; и далее использовать ее. Еще плюс в этом то, что избавит вас от дублирования кода, помимо того, что сделает ретёрн рендера более чистым и лучше читабельным.

Zaytsev-Alex commented on src/components/app-header/app-header.js:
```diff
@@ -0,0 +1,35 @@
+class AppHeader extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = { current: 1 };
```
Можно лучше: переменную можно переименовать, чтобы название лучше описывала предназначение, к примеру currentTab / currentPage.

Zaytsev-Alex commented on src/components/app-header/app-header.js:
```diff
@@ -0,0 +1,35 @@
+                        <Logo/>
+                    </div>
+                    <nav className={styles.nav}>
+                        <NavItem title="Конструктор"value={1} active={this.state.current===1} onClick={this.setCurrent}><BurgerIcon type={this.state.current===1?"primary":'secondary'} /></NavItem>
```
Можно лучше: value={1} не самый удачный вариант задать значение. В этом случае 1 является магическим числом (https://ru.hexlet.io/blog/posts/magic-numbers). Их лучше избегать, это позволяет увеличить надежность кода и улучшить его поддержку. Так же бывают и магические строки, они аналогичны числам, только строки :)
К примеру можно создать объект с возможными страницами:
const PAGES = {
CONSTRUCTOR: 1,
ORDERS: 2,
и тд
}
и далее использовать вместо 1 PAGES.CONSTRUCTOR

Zaytsev-Alex commented on src/components/burger-category/burger-category.js:
```diff
@@ -0,0 +1,26 @@
+    constructor(props) {
+        super(props);
+        this.state = {  }
+    }
```
Можно лучше: нет нужды вызывать конструктор и создавать в нем стейт, если он не используется. Конструктор вместе с инициализацией стейта можно упразднить.

Zaytsev-Alex commented on src/components/burger-category/burger-category.js:
```diff
@@ -0,0 +1,26 @@
+    }
+}
+ 
+export default BurgerCategory;
```
Нужно исправить: Пропущена валидация пропсов через propTypes. 
https://ru.reactjs.org/docs/typechecking-with-proptypes.html

Zaytsev-Alex commented on src/components/burger-category/burger-category.js:
```diff
@@ -0,0 +1,26 @@
+            <>
+                <p id={this.props.id} className={`text text_type_main-medium pt-10`}>{this.props.title}</p>
+                <ul className={styles.list}>
+                    { this.props.ingredients.map((ingredient) => {
+                        return( <BurgerIngredient name={ingredient.name} price={ingredient.price} image={ingredient.image}/> )
+                    }
```
Нужно исправить: при рендере списка следует использовать ключи (key).
https://ru.reactjs.org/docs/lists-and-keys.html
Они позволяют реакту лучше следить за состоянием и необходимостью перерисовывать списки.
В качестве ключа можно использовать поле _id объекта ингредиента.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -0,0 +1,48 @@
+        return ( 
+            <section>
+                <ul className={`mt-25 pl-4 pr-2 ${styles.container}`}>
+                    {this.state.ingredients.map((ingredient) => {
+                        return (
+                            <li className={`${styles.item} ${ingredient.isLocked?'pl-8':''}`}>
```
Нужно исправить: пропущен атрибут key при рендере списка. Подробнее описал в предыдущем комментарии.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -0,0 +1,48 @@
+                    {this.state.ingredients.map((ingredient) => {
+                        return (
+                            <li className={`${styles.item} ${ingredient.isLocked?'pl-8':''}`}>
+                                {!ingredient.isLocked? <DragIcon type="primary"/>:''}
```
Можно лучше: в качестве условного рендера, где рендерится либо элемент, либо ничего удобно использовать оператор &&.
Запись можно заменить на:
{!ingredient.isLocked && <DragIcon type="primary"/>}

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -0,0 +1,48 @@
+                        return (
+                            <li className={`${styles.item} ${ingredient.isLocked?'pl-8':''}`}>
+                                {!ingredient.isLocked? <DragIcon type="primary"/>:''}
+                                <ConstructorElement isLocked={ingredient.isLocked} type={ingredient.type} className='ml-2' text={ingredient.text} price={ingredient.price} thumbnail={ingredient.thumbnail}/>
```
Нужно исправить: по макету скролл должны быть по элементам, которые напрямую находятся в бургере, то есть между верхней и нижней булкой. Должен быть контейнер, в котором вверху и внизу находятся булки, между ними дополнительный контейнер со всеми остальными ингредиентами и скролл только во внутреннем.

Zaytsev-Alex commented on src/components/burger-ingredient/burger-ingredient.js:
```diff
@@ -0,0 +1,27 @@
+    render() { 
+        return (  
+            <li className={styles.ingredient}>
+                {this.state.count!==0? <Counter count={this.state.count} size="default"/>:<></>}
```
Можно лучше: аналогично можно использовать сокращенный условный рендер через &&. Так же, не стоит возвращать фрагмент(<></>) в случае, если ничего не возвращается. можно использовать null, это более общепринятая практика.

Zaytsev-Alex commented on src/components/burger-ingredient/burger-ingredient.js:
```diff
@@ -0,0 +1,27 @@
+    }
+}
+ 
+export default BurgerIngredient ;
```
Нужно исправить: пропущена типизация пропсов, необходимо использовать propTypes.

Zaytsev-Alex commented on src/components/burger-ingredient/burger-ingredient.js:
```diff
@@ -0,0 +1,27 @@
+import React from 'react'
+import styles from './burger-ingredient.module.css'
+
+class BurgerIngredient  extends React.Component {
```
Отлично: что ингредиент вынесен в отдельный файл.

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -0,0 +1,53 @@
+    state = {current:'Булки',main:[],sauce:[],bun:[]};
+    constructor(props) {
+        super(props);
+        data.map((ingredient) => {
```
Можно лучше: для простой итерации по массиву следует использовать метод forEach (https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
Метод map используется для создания нового массив основываясь на исходном.

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -0,0 +1,53 @@
+    constructor(props) {
+        super(props);
+        data.map((ingredient) => {
+            if(ingredient.type==='main'){
+                this.state = ({...this.state,main:[...this.state.main,ingredient]})
+            }
+            if(ingredient.type==='bun'){
+                this.state = ({...this.state,bun:[...this.state.bun,ingredient]})
+            }
+            if(ingredient.type==='sauce'){
+                this.state = ({...this.state,sauce:[...this.state.sauce,ingredient]})
+            }
```
Можно лучше: стоит сначала, к примеру, в отдельной константе создать стейт используя данные и потом записать его в this.state вместо переопределения стейта. Или же можно сделать что-то такое:
this.state = {
bun: data.filter((ingredient) => ingredient.type==='bun'
и тд
}

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -0,0 +1,53 @@
+        })
+    }
+    
+     setCurrent = (value) => {
+         this.setState({...this.state,current:value})
+    }
```
Отлично: круто, что на самом раннем этапе уже реализована навигации по ингредиентам!

Zaytsev-Alex commented on src/components/nav-item/nav-item.js:
```diff
@@ -0,0 +1,21 @@
+    
+    render() { 
+        return (
+            <div onClick={()=>{this.props.onClick(this.props.value)}} className={`mt-4 mb-4 pb-4 pt-4 pr-5 pl-5 ${styles.container}`}>
```
Можно лучше: метод onClick лучше обработать в методе компонента NavItem. В теле компонента можно создать функцию handleNavItemClick и в ней реализовать логику, чтобы избежать инлайнового создания функции в рендере (она при каждом ререндере будет менять ссылку и компонент куда она передается будет в любом случае перерендериваться). Так же это позволяет держать рендер компонента в чистоте.
Дополнительно для навигации пока лучше использовать ссылки <a/>

Zaytsev-Alex commented on src/components/nav-item/nav-item.js:
```diff
@@ -0,0 +1,21 @@
+    }
+}
+ 
+export default NavItem;
```
Нужно исправить: пропущена типизация через propTypes.

Zaytsev-Alex commented on src/components/nav-item/nav-item.js:
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import styles from './nav-item.module.css'
+
+class NavItem extends React.Component {
```
Отлично: круто, что применена декомпозиция и создан отдельный компонент.

Zaytsev-Alex commented on src/components/burger-ingredient/burger-ingredient.js:
```diff
@@ -0,0 +1,27 @@
+        return (  
+            <li className={styles.ingredient}>
+                {this.state.count!==0? <Counter count={this.state.count} size="default"/>:<></>}
+                <img src={this.props.image} alt=''/>
```
Можно лучше: добавление альтернативного текста для картинки позволит увеличить доступность. можно использовать название ингридиента.

Zaytsev-Alex reviewed (commented):
Отличная работа, практически полностью совпадает с макетом. Круто, что применена декомпозиция компонентов и созданы вспомогательные компоненты и, что уже на такой ранней стадии можно переключаться по категориями ингридиентов.

Из того, что нужно исправить:
Компонент app вместе с его стилями должен быть помещен в папку компонентов. Также стоит его назвать с маленькой буквы, чтобы поддерживаться наименования в kebab-case. Еще для app лучше использовать jsx расширение, т.к. он не использует typescript.
Следует использовать css-modules вместо обычного подключения стилей или инлайнового объявления.
При рендере списков использовать key.
Типизировать пропсы через propTypes.
Скролл по ингредиентам бургера должен быть между булок (подробнее описано в соответствующем месте в коде). Так же над и по бокам от хедера не должно быть отступов, как и вертикального скролла на основной странице. Скролл должен быть реализован в соответствующих секциях.

Больше комментариев по улучшению кода можно найти в основных комментариях.
Удачи в дальнейших работах!

DemasX64 committed:
add props validation
remove scroll on page
remove inline css/add css modules
fix scroll in constructor

more fixes for performance & readability
